app/lib/boolean-phrase-parser.js

```diff
@@ -0,0 +1,208 @@
+            throw new Error('can not evaluate an empty expression');
+        }
+
+        stringExpression = stringExpression.trim();
+
+        var removedAllParentheses = false;
+        var result;
+        while (!removedAllParentheses) {
+            result = this.analyzeExpression(stringExpression);
+            if (result.parenthesesExpression) {
+                //peal off the parentheses
+                stringExpression = stringExpression.substring(1, stringExpression.length - 1);
+            } else {
+                removedAllParentheses = true;
+            }
+        }
+
+        if (result.unaryOperator) {
+            var childSubTree = new ExpressionNode(result.expression, this.depth + 1);
+            this.data = this.createUnaryNotNodeData(childSubTree);
+        } else if (result.literal) {
+            this.data = this.createLiteralNodeData(result.literal);
+        } else if (result.binaryOperator) {
+            this.data = this.createBinaryOperandNodeData(
+                new ExpressionNode(result.leftOperand, this.depth + 1),
+                new ExpressionNode(result.rightOperand, this.depth + 1),
+                result.binaryOperator);
+        } else {
+            throw new Error('this should not have happened');
+        }
+    }
+
+    ExpressionNode.prototype.analyzeExpression = function (expression) {
+        var openParentheses = 0;
+        var closeParentheses = 0;
+        var binaryOperator = null;
+        var leftOperatorEndIndex = 0;
+        var currentChar;
+
+        for (var i = 0, length = expression.length; i < length; i++) {
+            currentChar = expression[i];
+
+            if (currentChar === openParethesisChar) {
+                openParentheses++;
+            } else if (currentChar === closeParethesisChar) {
+                closeParentheses++;
+            }
+
+            if (this.isWhiteSpaceChar(currentChar)) {
+                continue;
+            }
+
+            if (openParentheses > closeParentheses) {
+                continue;
+            } else if ((openParentheses === closeParentheses) && this.isBinaryOperand(currentChar)) {
+                //there are two cases:
+                //(1) if we don't have a binaryOperator yet save the currentChar as the one
+                //(2) if we do have one already and it's AND and the current one is OR prefer it (this creates the AND>OR operator precedence)
+                //    the AND will be left to be created as one of the children in the tree and thus evaluated first.
+                if ((!binaryOperator) ||
+                    (binaryOperator && binaryOperator === andChar && currentChar === orChar)) {
+                    binaryOperator = currentChar;
+                    leftOperatorEndIndex = i;
+                }
+
+            } else if (closeParentheses > openParentheses) {
+                //TODO: add some more information here
+                throw new Error('mismatch of parentheses');
```

Comment:
This comment is for all the thrown exceptions - please provide any information you have to where this happened (e.g. the expression)